pkg/diskmaker/diskmaker.go

```diff
@@ -125,27 +131,85 @@ func (d *DiskMaker) symLinkDisks(diskConfig DiskConfig) {
 
 }
 
-func (d *DiskMaker) findMatchingDisks(diskConfig DiskConfig, deviceSet sets.String) (map[string][]string, error) {
-	blockDeviceMap := make(map[string][]string)
-	addDiskToMap := func(scName, diskName string) {
+func (d *DiskMaker) findMatchingDisks(diskConfig DiskConfig, deviceSet sets.String) (map[string][]DiskLocation, error) {
+	blockDeviceMap := make(map[string][]DiskLocation)
+	addDiskToMap := func(scName, stableDeviceID, diskName string) {
 		deviceArray, ok := blockDeviceMap[scName]
 		if !ok {
-			deviceArray = []string{}
+			deviceArray = []DiskLocation{}
 		}
-		deviceArray = append(deviceArray, diskName)
+		deviceArray = append(deviceArray, DiskLocation{diskName, stableDeviceID})
 		blockDeviceMap[scName] = deviceArray
 	}
-	for blockDevice := range deviceSet {
-		for storageClass, disks := range diskConfig {
-			if hasExactDisk(disks.DiskNames, blockDevice) {
-				addDiskToMap(storageClass, blockDevice)
-				break
+	for storageClass, disks := range diskConfig {
+		// handle diskNames
+		for _, diskName := range disks.DiskNames {
+			if hasExactDisk(deviceSet, diskName) {
+				devicePath, err := d.findStableDeviceID(diskName)
+				if err != nil {
+					logrus.Errorf("Unable to add disk %s to local disk pool %v", diskName, err)
+					continue
+				}
+				addDiskToMap(storageClass, devicePath, diskName)
+				continue
+			}
+		}
+		// handle DeviceIDs
+		for _, deviceID := range disks.DeviceIDs {
+			diskPath, deviceName, err := d.findDeviceByID(deviceID)
+			if err != nil {
+				logrus.Errorf("unable to add disk %s to local disk pool %v", deviceID, err)
+				continue
 			}
+			addDiskToMap(storageClass, diskPath, deviceName)
 		}
 	}
 	return blockDeviceMap, nil
 }
 
+// findDeviceByID finds device ID and return name and path
+func (d *DiskMaker) findDeviceByID(deviceID string) (string, string, error) {
```

Comment:
If I understand it correctly (and I am not sure because the code is confusing), this function gets stable device (XXX from `/dev/disk/by-id/XXX`) and finds its corresponding `/dev/YYY`. Why can't you just do `EvalSymlinks(XXX)` and scan whole /dev/disk/by-id instead?

Reply:
yeah that is simpler. I will change it.